students/shawn_degraw/session07/html_render.py

```diff
@@ -0,0 +1,159 @@
+#!/usr/bin/env python3
+
+"""
+A class-based system for rendering html.
+"""
+
+
+class Element(object):
+    """ Default Element class for rendering html """
+    tag_name = "html"
+    indent = "   "
+    cur_ind = ""
+
+    def __init__(self, content=None, **kwargs):
+        self.html_content = ([content] if content else [])
+        self.header_arg = (kwargs if kwargs else {})
+
+    def append(self, new_content):
+        self.html_content.append(new_content)
+
+    def render(self, out_file, cur_ind=""):
+        if self.header_arg:
+            self.writetag(out_file, cur_ind + "<{}")
+            for k, v in self.header_arg.items():
+                out_file.write(" {}=\"{}\"".format(k, v))
+            out_file.write(">\n")
+        else:
+            self.writetag(out_file, cur_ind + "<{}>\n")
+
+        for htmlitem in self.html_content:
+            if type(htmlitem) == str:
+                out_file.write((cur_ind + self.indent) + "{}\n".format(htmlitem))
+            else:
+                htmlitem.render(out_file, (cur_ind + self.indent))
+
+        self.writetag(out_file, cur_ind + "</{}>\n")
+
+    def writetag(self, out_file, outstring):
+        out_file.write(outstring.format(self.tag_name))
+
+
+class OneLineTag(Element):
+    """ subclass of Element for rendering single line html tags """
+    tag_name = ""
+
+    def render(self, out_file, cur_ind=""):
+        if self.header_arg:
+            self.writetag(out_file, cur_ind + "<{}")
+            for k, v in self.header_arg.items():
+                out_file.write(" {}=\"{}\"".format(k, v))
+            out_file.write(">")
+        else:
+            self.writetag(out_file, cur_ind + "<{}>")
+
+        for htmlitem in self.html_content:
+            out_file.write("{}".format(htmlitem))
+
+        self.writetag(out_file, "</{}>\n")
+
+
+class SelfClosingTag(Element):
+    """ subclass of Element for rendering tags with no content """
+    tag_name = ""
+
+    def __init__(self, content=None, **kwargs):
+        if content:
+            raise TypeError('Error: Content not allowed for tag.')
+        else:
+            self.html_content = []
+
+        self.header_arg = (kwargs if kwargs else {})
```

Comment:
You are not reusing the superclass initializer, but redefining the `self.html_content` and `self.header_arg` here separately. It's better to use `super().__init__(content=content, kwargs=kwargs)`. You can see this instruction in the tutorial. Maybe I didn't emphasize how to invoke the superclass initializer from a subclass...

Reply:
I think I have this corrected in the newest commit.